spring-kafka/src/main/java/org/springframework/kafka/requestreply/AggregatingReplyingKafkaTemplate.java

```diff
@@ -0,0 +1,281 @@
+		Assert.isTrue(ackMode.equals(AckMode.MANUAL) || ackMode.equals(AckMode.MANUAL_IMMEDIATE),
+				"The reply container must have a MANUAL or MANUAL_IMMEDIATE AckMode");
+		this.releaseStrategy = releaseStrategy;
+	}
+
+	/**
+	 * Set the timeout to use when committing offsets.
+	 * @param commitTimeout the timeout.
+	 */
+	public void setCommitTimeout(Duration commitTimeout) {
+		Assert.notNull(commitTimeout, "'commitTimeout' cannot be null");
+		this.commitTimeout = commitTimeout;
+	}
+
+	/**
+	 * Set to true to return a partial result when a request times out.
+	 * @param returnPartialOnTimeout true to return a partial result.
+	 */
+	public void setReturnPartialOnTimeout(boolean returnPartialOnTimeout) {
+		this.returnPartialOnTimeout = returnPartialOnTimeout;
+	}
+
+	@Override
+	public void onMessage(List<ConsumerRecord<K, Collection<ConsumerRecord<K, R>>>> data, Consumer<?, ?> consumer) {
+		long now = System.currentTimeMillis();
+		if (now - this.lastOrphanCheck > getReplyTimeout() * 10) { // NOSONAR magic #
+			weedOrphans(consumer);
+		}
+		List<ConsumerRecord<K, Collection<ConsumerRecord<K, R>>>> completed = new ArrayList<>();
+		data.forEach(record -> {
+			Header correlation = record.headers().lastHeader(KafkaHeaders.CORRELATION_ID);
+			if (correlation == null) {
+				this.logger.error("No correlationId found in reply: " + record
+						+ " - to use request/reply semantics, the responding server must return the correlation id "
+						+ " in the '" + KafkaHeaders.CORRELATION_ID + "' header");
+			}
+			else {
+				CorrelationKey correlationId = new CorrelationKey(correlation.value());
+				List<ConsumerRecord<K, R>> list = addToCollection(record, correlationId, now).stream()
+						.map(entry -> entry.getRecord())
+						.collect(Collectors.toList());
+				if (isPending(correlationId) && this.releaseStrategy.test(list)) {
+					ConsumerRecord<K, Collection<ConsumerRecord<K, R>>> done =
+							new ConsumerRecord<>(AGGREGATED_RESULTS_TOPIC, 0, 0L, null, list);
+					done.headers().add(new RecordHeader(KafkaHeaders.CORRELATION_ID, correlationId.getCorrelationId()));
+					this.pending.remove(correlationId);
+					checkOffsets(list);
+					commitIfNecessary(consumer);
+					completed.add(done);
+				}
+			}
+		});
+		if (completed.size() > 0) {
+			super.onMessage(completed);
+		}
+	}
+
+	/**
+	 * We may receive a partial delivery of a previously completed group of replies
+	 * e.g. after a rebalance. To avoid a memory leak, check for such conditions and
+	 * discard from time-to-time.
+	 * @param consumer the consumer.
+	 */
+	private void weedOrphans(Consumer<?, ?> consumer) {
+		Map<CorrelationKey, List<ConsumerRecord<K, R>>> orphaned = this.pending.entrySet()
+			.stream()
+			.filter(entry -> entry.getValue()
+					.stream()
+					.allMatch(holder -> holder.getTimestamp() > this.lastOrphanCheck))
+			.collect(Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue()
```

Comment:
Looking into this one more time, I would say we don't need to collect at all: there is just enough to *filter* and then `forEach`, without intermediate `List` and `Map`.

Also it might be even better to rethink this Java Streams usage and just use an `Iterator` from `this..pending.entrySet()`

Reply:
I collected them so I could log the discarded records. Or do you mean log them one correlation key at-a-time?

How would I do the `allMatch()` test if I use a simple iterator? I'd have to iterate over each list.

Reply:
Pushed the other changes while we discuss this further.

Reply:
> Pushed the other changes

Looks like not yet.

OK. If the logging is the point, then never mind. Let's live with collecting!

Reply:
Sorry; committed but didn't push.